### PR TITLE
Added checks for downloaded source and wheel file

### PR DESCRIPTION
**Pull request by rd4398**
Added the checks for downloaded source and wheel files by validating file extensions and matching requirement with downloaded tar

### CODE REVIEW

File: src/fromager/sdist.py [hunk 1]
```diff
@@ -232,6 +233,15 @@ def handle_requirement(
 
     return resolved_version
 
+# Helper method to check whether the .whl file is a zip file and has contents in it.
+# It will throw BadZipFile exception if any other file is encountered. Eg: index.html
+def download_wheel_check(wheels_prebuilt, wheel_url):
```
Comment (dhellmann) [inline src/fromager/sdist.py:238]: Without the `ctx` prefix in the caller I don't know what `wheels_prebuilt` means. That's really the directory where the file is supposed to be downloaded, right? What is the similar argument called in `download_url()`?

File: src/fromager/sources.py [hunk 2]
```diff
@@ -129,12 +129,29 @@ def default_download_source(
...
-    source_filename = download_url(ctx.sdists_downloads, url)
+    source_filename = download_source_check(ctx.sdists_downloads, url)
     logger.debug(
         f"{req.name}: have source for {req} version {version} in {source_filename}"
     )
     return (source_filename, version, url)
 
+# Helper method to check whether .zip /.tar / .tgz is able to extract and check its content.
+# It will throw exception if any other file is encountered. Eg: index.html
+def download_source_check(sdists_downloads, url):
```
Comment (dhellmann) [inline src/fromager/sources.py:140]: Same comment on the name of the first argument here.

File: src/fromager/sdist.py [hunk 3]
```diff
@@ -232,6 +233,15 @@ def handle_requirement(
...
+    wheel_filename = sources.download_url(wheels_prebuilt, wheel_url)
+    wheel_directory_contents = ZipFile(wheel_filename).namelist()
+    if len(wheel_directory_contents) == 0:
```
Comment (dhellmann) [inline src/fromager/sdist.py:241]:
```suggestion
    if not wheel_directory_contents:
```

File: src/fromager/sources.py [hunk 4]
```diff
@@ -129,12 +129,29 @@ def default_download_source(
...
+    source_filename = download_url(sdists_downloads, url)
+    if source_filename.suffix == ".zip":
+        source_file_contents = ZipFile(source_filename).namelist()
+        if len(source_file_contents) == 0:
```
Comment (dhellmann) [inline src/fromager/sources.py:144]:
```suggestion
        if not source_file_contents:
```

File: src/fromager/sources.py [hunk 5]
```diff
@@ -129,12 +129,29 @@ def default_download_source(
...
+            raise TypeError("Bad Zip File encountered")
+    elif source_filename.suffix == ".tgz" or source_filename.suffix == ".gz":
+        tar = tarfile.open(source_filename)
+        contents = tar.getnames()
+        tar.close()
+        if len(contents) == 0:
```
Comment (dhellmann) [inline src/fromager/sources.py:150]:
```suggestion
        if not contents:
```

File: src/fromager/sources.py [hunk 6]
```diff
@@ -129,12 +129,29 @@ def default_download_source(
...
+# Helper method to check whether .zip /.tar / .tgz is able to extract and check its content.
+# It will throw exception if any other file is encountered. Eg: index.html
+def download_source_check(sdists_downloads, url):
+    source_filename = download_url(sdists_downloads, url)
+    if source_filename.suffix == ".zip":
+        source_file_contents = ZipFile(source_filename).namelist()
+        if len(source_file_contents) == 0:
+            raise TypeError("Bad Zip File encountered")
+    elif source_filename.suffix == ".tgz" or source_filename.suffix == ".gz":
+        tar = tarfile.open(source_filename)
```
Comment (dhellmann) [inline src/fromager/sources.py:147]: If you use `with` you don't have to close the file explicitly.

```python
with tarfile.open(source_filename) as tar:
  contents = tar.getnames()
```

File: tests/test_sdist.py [hunk 7]
```diff
@@ -30,3 +31,13 @@ def test_missing_dependency_format(resolve_dist, tmp_context):
     assert "setuptools>=40.8.0 -> 69.5.1" in s
     # Ensure we report what version we expect of all of the other dependencies
     assert "flit_core -> 3.9.0" in s
+
+@patch("fromager.sources.download_url")
+@patch("zipfile.ZipFile")
+def test_invalid_wheel_file_exception(mock_zipfile, mock_download_url):
+    mock_download_url.return_value = "fake_wheel.csv"
```
Comment (dhellmann) [inline tests/test_sdist.py:38]: In order for the new check function to work, this file has to actually exist somewhere on the filesystem.

File: tests/test_sdist.py [hunk 8]
```diff
@@ -30,3 +31,13 @@ def test_missing_dependency_format(resolve_dist, tmp_context):
...
+    mock_zip_instance = MagicMock()
+    mock_zip_instance.namelist.return_value = ["file1.txt", "file2.txt"]
```
Comment (dhellmann) [inline tests/test_sdist.py:40]: Mocking out the `namelist()` method like this bypasses the checks in your test function. A more robust test would be to create the file on the filesystem as something other than a zip file, then let ZipFile try to read that file.

File: tests/test_sdist.py [hunk 9]
```diff
@@ -30,3 +31,13 @@ def test_missing_dependency_format(resolve_dist, tmp_context):
...
+    mock_zipfile.return_value = mock_zip_instance
+    sdist.download_wheel_check(["prebuilt1", "prebuilt2"], "http://example.com/fake_wheel.csv")
+    assertRaises(exceptions.TypeError, sdist.download_wheel_check)
```
Comment (dhellmann) [inline tests/test_sdist.py:43]: I think you want to be using `with pytest.raises()` here. There's an example in https://github.com/python-wheel-build/fromager/blob/main/tests/test_constraints.py#L25

**Review by dhellmann — CHANGES_REQUESTED** [review]

File: src/fromager/sdist.py (hunk 1 above)
Comment (rd4398) [reply src/fromager/sdist.py:238]: If I try to do `ctx.wheels_prebuilt`, it gives me syntax error for arguments. Should I pass `ctx` as parameter? or create a new `ctx` in helper function?

Comment (dhellmann) [reply src/fromager/sdist.py:238]: Sorry, I wasn't clear. The variable name doesn't make sense. Can we rename it to `destination_dir` like the argument to `download_url()`?

Comment (rd4398) [reply src/fromager/sdist.py:238]: oh, okay! sure I will do that!

**Pull request closed without merging**